R/check_forecasts.R

```diff
@@ -0,0 +1,267 @@
+                  paste(check[["forecast_unit"]], collapse = "`, `"), "`. ",
+                  "If this is not as intended, please DELETE UNNECESSARY columns or add new ones."))
+
+  # check what format is has right now and tell user to convert it.
+  if (!any(c("quantile", "sample") %in% colnames(data))) {
+    if ("range" %in% colnames(data) | any(grepl("lower_", colnames(data)))) {
+      errors <- c(errors,
+                  "It seems like you have a format based on forecast intervals (see `example_data_long`, `example_data_semi_wide`, `example_data_wide`). You need to convert this to a quantile-based format first using `range_wide_to_long()` and `range_long_to_quantile()`")
+
+    } else if (!check[["target_type"]] == "binary") {
+      errors <- c(errors,
+                  "This forecast does not seem to be for a binary prediction target, so we need a column called quantile or sample")
+    }
+  }
+
+  # check whether there is more than one prediction for the same target, i.e.
+  # the length of prediction is greater 1 for a sample / quantile for
+  # a single forecast
+  type <- c("sample", "quantile")[c("sample", "quantile") %in% colnames(data)]
+  data[, InternalDuplicateCheck := .N, by = c(obs_unit, type)]
+
+  if (any(data$duplicatecheck) > 1) {
+    errors <- c(errors,
+                paste("There are instances with more than one forecast for the same target.",
+                      "This can't be right and needs to be resolved. Maybe you need to check",
+                      "the unit of a single forecast and add missing columns?"))
+    check[["duplicate_forecasts"]] <- data[InternalDuplicateCheck > 1]
+  }
+  data[, InternalDuplicateCheck := NULL]
+
+  # check whether there is a model column present. And if not, state what that means
+  if (!("model" %in% colnames(data))) {
+    msg <- c(msg,
+             paste("There is no column called `model` in the data.",
+                   "scoringutils therefore thinks that all forecasts come from the same model"))
+    data[, model := "Unspecified model"]
+
+  }
+
+  # some checks whether there are the same number of quantiles, samples
+  data[, InternalNumCheck := length(prediction), by = obs_unit]
+  n <- unique(data$InternalNumCheck)
+  if (length(n) > 1) {
+    warnings <- c(warnings,
+                  paste0("Some forecasts have different numbers of rows (e.g. quantiles or samples). ",
+                         "scoringutils found: ", paste(n, collapse = ", "),
+                         ". This is not necessarily a problem, but make sure this is intended."))
+  }
+  check[["rows_per_forecast"]] <-
+    data[, .(rows_per_forecast = unique(InternalNumCheck)), by = model]
+  data[, InternalNumCheck := NULL]
+
+  # get available unique values per model for the different columns
+  cols <- obs_unit[!(obs_unit == "model")]
+  check[["unique_values"]] <-
+    data[, lapply(.SD, FUN = function(x) length(unique(x))), by = "model"]
```

Comment:
```suggestion
    data[, vapply(.SD, FUN = function(x) length(unique(x)), integer(1)), by = "model"]
```

Reply:
The second version is faster?

Reply:
Slightly faster and generally considered good practice since you can control the output (equivalent to `purrr::map_dbl()` vs `purrr::map()`)